### PR TITLE
GridPlus integration

### PR DESCRIPTION
**Pull request by FrederikBolding**

### CODE REVIEW

File: src/assets/images/wallets/gridplus.svg [hunk 1]
```diff
@@ -0,0 +1 @@
...data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 656.49 290"><defs><style>.cls-1{fill:#41b6e6;}</style></defs><title> grid-plus-logo-black+blue</title><polygon class="cls-1" points="590.89 72.05 558.48 72.05 558.48 40.17 557.93 40.17 533.88 40.17 531.4 40.17 531.4 72.05 498.99 72.05 498.99 97.58 531.4 97.58 531.4 129.46 558.48 129.46 558.48 97.58 590.89 97.58 590.89 72.05"/><path d="M169.59,188.13c-6.7,3.24-15.62,5.47-26.37,5.47-27,0-48.87-19.47-48.87-47.86s20.88-47.05,49.07-47.05c19.68,0,31.64,7.3,35.29,9.73l12.37-22.31c-5.27-3.44-20.28-13.58-48.87-13.58-45,0-76.87,32-76.87,73.21s31.84,73,76.26,73c29.81,0,46.65-10.75,54.35-16.63V145.94H169.59Z"/><path d="M336.4,125c0-32.76-23.8-53-59.48-53H224.53V218.75h28.8V176.8h25.26l28.59,41.95h34l-33.6-48.42C325.75,161.77,336.4,145.91,336.4,125Zm-62,26.51H253.33V97.29h21.29c19,0,31.93,10.22,31.93,27.75C306.55,141.11,293.82,151.55,274.41,151.55Z"/><rect x="366.35" y="72.03" width="28.8" height="146.72"/><path d="M558.33,146.62h0v-4.41H531.47v4.41s-.29,10.68-.76,14.2c-2.75,20.52-22.23,32.47-47.11,32.47H457.09V97.49h26.65V72H428.29V218.75h56.56c40.64,0,70.27-22.21,73.48-57.93C558.65,157.29,558.33,146.62,558.33,146.62Z"/></svg>
```
Comment (Mrtenz) [inline src/assets/images/wallets/gridplus.svg:1]: Add new line

File: src/components/SignTransactionWallets/Hardware.tsx [hunk 2]
```diff
@@ -94,15 +99,18 @@ export default function HardwareSignTransaction({
...
         wallet
           .signTransaction(madeTx)
           .then((data) => {
             // User approves tx.
             setIsTxSignatureRequestDenied(false);
+            setIsRequestingTxSignature(false);
             onSuccess(data);
           })
-          .catch(() => {
+          .catch((err) => {
```
Comment (Mrtenz) [inline src/components/SignTransactionWallets/Hardware.tsx:112]: Can this be tested

File: src/components/WalletUnlock/GridPlus.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    mergedDPaths
+  } = useHDWallet(extendedDPaths, WalletId.GRIDPLUS, DEFAULT_GAP_TO_SCAN_FOR);
+
+  const handleNullConnect = () => {
+    requestConnection(network, assetToUse);
+  };
+
+  const handleAssetUpdate = (newAsset: ExtendedAsset) => {
+    setAssetToUse(newAsset);
+    updateAsset(newAsset);
```
Comment (Mrtenz) [inline src/components/WalletUnlock/GridPlus.tsx:66]: Can this be tested

File: src/components/WalletUnlock/GridPlus.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  useHDWallet,
+  useNetworks
+} from '@services';
+import { ExtendedAsset, FormData, WalletId } from '@types';
+
+import HardwareWalletUI from './Hardware';
+
+interface OwnProps {
+  formData: FormData;
+  onUnlock(param: any): void;
```
Comment (Mrtenz) [inline src/components/WalletUnlock/GridPlus.tsx:23]: Don't use `any`

File: src/components/WalletUnlock/GridPlus.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+export const GridPlus = ({ formData, onUnlock }: OwnProps) => {
+  const { networks } = useNetworks();
+  const { assets } = useAssets();
+  const network = getNetworkById(formData.network, networks);
+  const baseAsset = getAssetByUUID(assets)(network.baseAsset) as ExtendedAsset;
+  const dpaths = uniqBy(prop('path'), [
+    ...getDPaths([network], WalletId.GRIDPLUS),
+    ...GRIDPLUS_DERIVATION_PATHS
+  ]);
```
Comment (Mrtenz) [inline src/components/WalletUnlock/GridPlus.tsx:34]: Can you do this without ramda

File: src/components/WalletUnlock/GridPlus.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        dpaths={mergedDPaths}
+        setSelectedDPath={setSelectedDPath}
+        updateAsset={updateAsset}
+        addDPaths={addDPaths}
+        scanMoreAddresses={scanMoreAddresses}
+        handleAssetUpdate={handleAssetUpdate}
+        onUnlock={onUnlock}
+      />
+    );
+  } else {
```
Comment (Mrtenz) [inline src/components/WalletUnlock/GridPlus.tsx:88]: Unnecessary `else`

File: src/components/WalletUnlock/GridPlus.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+export const GridPlus = ({ formData, onUnlock }: OwnProps) => {
+  const { networks } = useNetworks();
+  const { assets } = useAssets();
+  const network = getNetworkById(formData.network, networks);
+  const baseAsset = getAssetByUUID(assets)(network.baseAsset) as ExtendedAsset;
+  const dpaths = uniqBy(prop('path'), [
+    ...getDPaths([network], WalletId.GRIDPLUS),
+    ...GRIDPLUS_DERIVATION_PATHS
+  ]);
+  const defaultDPath = network.dPaths[WalletId.GRIDPLUS] || DEFAULT_ETH;
```
Comment (Mrtenz) [inline src/components/WalletUnlock/GridPlus.tsx:35]:
```suggestion
  const defaultDPath = network.dPaths[WalletId.GRIDPLUS] ?? DEFAULT_ETH;
```

File: src/components/WalletUnlock/HWLegacy.tsx [hunk 8]
```diff
@@ -37,19 +43,17 @@ interface State {
...
-class LedgerNanoSDecryptClass extends PureComponent<Props & INetworkContext, State> {
+class HWLegacyClass extends PureComponent<Props & INetworkContext, State> {
   public state: State = {
     dPath:
-      getDPath(this.props.getNetworkById(this.props.formData.network), WalletId.LEDGER_NANO_S) ||
-      getDPaths(this.props.networks, WalletId.LEDGER_NANO_S)[0],
+      getDPath(
+        this.props.getNetworkById(this.props.formData.network),
+        this.props.wallet.id as DPathFormat
+      ) || getDPaths(this.props.networks, this.props.wallet.id as DPathFormat)[0],
```
Comment (Mrtenz) [inline src/components/WalletUnlock/HWLegacy.tsx:52]:
```suggestion
      ) ?? getDPaths(this.props.networks, this.props.wallet.id as DPathFormat)[0],
```

File: src/components/WalletUnlock/HWLegacy.tsx [hunk 9]
```diff
@@ -177,9 +203,16 @@ class LedgerNanoSDecryptClass extends PureComponent<Props & INetworkContext, Sta
...
 }
 
-export const LedgerNanoSDecrypt = withHook(useNetworks)(LedgerNanoSDecryptClass);
+const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) =>
+  bindActionCreators({ connectWallet }, dispatch);
+
+const connector = connect(() => ({}), mapDispatchToProps);
+type Props = ConnectedProps<typeof connector> & OwnProps;
+
+export const HWLegacy = withHook(useNetworks)(connector(HWLegacyClass));
```
Comment (Mrtenz) [inline src/components/WalletUnlock/HWLegacy.tsx:218]: Can you test this file?

File: src/config/wallets.ts [hunk 10]
```diff
@@ -257,3 +273,60 @@ export const WEB3_WALLETS: WalletSubType = filterObjectOfObjects(WALLETS_CONFIG)
...
+}
+
+type THardwareConfigs = {
+  [key in
+    | WalletId.LEDGER_NANO_S_NEW
+    | WalletId.TREZOR_NEW
+    | WalletId.GRIDPLUS
+    | WalletId.LEDGER_NANO_S
+    | WalletId.TREZOR]: HWConfig;
+};
```
Comment (Mrtenz) [inline src/config/wallets.ts:294]: `Record`?

File: src/features/AddAccount/AddAccountForm.reducer.tsx [hunk 11]
```diff
@@ -64,6 +64,7 @@ const handleUnlock = (walletType: WalletId | undefined, payload: any) => {
         ];
 
       case WalletId.TREZOR_NEW:
+      case WalletId.GRIDPLUS:
```
Comment (Mrtenz) [inline src/features/AddAccount/AddAccountForm.reducer.tsx:67]: Can you test this

File: src/services/Store/store/connections.slice.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * Sagas
+ */
+export function* connectionsSaga() {
+  yield all([takeLatest(connectWallet.type, gridPlusWorker)]);
+}
+
+export function* gridPlusWorker({ payload }: PayloadAction<Wallet>) {
+  const credentials =
+    payload && 'getCredentials' in payload && (payload as GridPlusWallet).getCredentials();
```
Comment (Mrtenz) [inline src/services/Store/store/connections.slice.ts:58]: Maybe use a type guard

File: src/services/WalletService/deterministic/useHDWallet.ts [hunk 13]
```diff
@@ -51,22 +51,22 @@ export const useHDWallet = (
...
       gap
     );
     if (newGapItems.length !== 0) {
-      dispatch(getAccounts({ session, dpaths: newGapItems }));
+      dispatch(getAccounts({ dpaths: newGapItems }));
       return;
     }
     if (customDPathItems.length > 0) {
-      dispatch(getAccounts({ session, dpaths: customDPathItems }));
+      dispatch(getAccounts({ dpaths: customDPathItems }));
```
Comment (Mrtenz) [inline src/services/WalletService/deterministic/useHDWallet.ts:69]: Can you test this

File: src/services/WalletService/deterministic/useHDWallet.ts [hunk 14]
```diff
@@ -89,11 +89,11 @@ export const useHDWallet = (
 
   const scanMoreAddresses = (dpath: ExtendedDPath) => {
     if (!isConnected || !network || !session) return;
-    dispatch(getAccounts({ session, dpaths: [dpath] }));
+    dispatch(getAccounts({ dpaths: [dpath] }));
```
Comment (Mrtenz) [inline src/services/WalletService/deterministic/useHDWallet.ts:92]: Can you test this

File: src/services/WalletService/walletService.ts [hunk 15]
```diff
@@ -37,20 +37,24 @@ export const WalletFactory = {
     init: ({ networks }: Web3WalletInitArgs) => unlockWeb3(networks)
   },
   [WalletId.LEDGER_NANO_S_NEW]: {
-    init: ({ address, dPath, index }: HardwareWalletInitArgs) =>
-      new LedgerWallet().getWallet(dPath, index, address)
+    init: ({ address, dPath, index, params }: HardwareWalletInitArgs) =>
+      getWallet(WalletId.LEDGER_NANO_S_NEW, params)!.getWallet(dPath, index, address)
```
Comment (Mrtenz) [inline src/services/WalletService/walletService.ts:41]: Can you test this

File: src/services/WalletService/walletService.ts [hunk 16]
```diff
@@ -37,20 +37,24 @@ export const WalletFactory = {
...
-    init: ({ address, dPath, index }: HardwareWalletInitArgs) =>
-      new LedgerWallet().getWallet(dPath, index, address)
+    init: ({ address, dPath, index, params }: HardwareWalletInitArgs) =>
+      getWallet(WalletId.LEDGER_NANO_S, params)!.getWallet(dPath, index, address)
   },
   [WalletId.TREZOR_NEW]: {
-    init: ({ address, dPath, index }: HardwareWalletInitArgs) =>
-      new TrezorWallet(trezorManifest).getWallet(dPath, index, address)
+    init: ({ address, dPath, index, params }: HardwareWalletInitArgs) =>
+      getWallet(WalletId.TREZOR_NEW, params)!.getWallet(dPath, index, address)
```
Comment (Mrtenz) [inline src/services/WalletService/walletService.ts:49]: Can you test this

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: src/components/WalletUnlock/GridPlus.tsx (hunk 5 above)
Comment (FrederikBolding) [reply src/components/WalletUnlock/GridPlus.tsx:34]: Yes, but would prefer not to 😄

File: src/config/wallets.ts (hunk 10 above)
Comment (FrederikBolding) [reply src/config/wallets.ts:294]: :+

File: src/features/AddAccount/AddAccountForm.reducer.tsx (hunk 11 above)
Comment (FrederikBolding) [reply src/features/AddAccount/AddAccountForm.reducer.tsx:65]: Not really easily 🤔

File: src/components/WalletUnlock/HWLegacy.tsx (hunk 9 above)
Comment (FrederikBolding) [reply src/components/WalletUnlock/HWLegacy.tsx:218]: I've tested most of the file. The rest is not easily testable at this point.

File: src/components/WalletUnlock/GridPlus.tsx (hunk 3 above)
Comment (FrederikBolding) [reply src/components/WalletUnlock/GridPlus.tsx]: Moved this to a shared file

File: src/services/WalletService/deterministic/useHDWallet.ts (hunk 13 above)
Comment (FrederikBolding) [reply src/services/WalletService/deterministic/useHDWallet.ts:69]: Not easily.

File: src/services/Store/store/connections.slice.ts (hunk 12 above)
Comment (FrederikBolding) [reply src/services/Store/store/connections.slice.ts:58]: On what condition? Just `'getCredentials' in payload`?

File: src/components/WalletUnlock/GridPlus.tsx (hunk 5 above)
Comment (Dqddyvillager1722) [reply src/components/WalletUnlock/GridPlus.tsx]: Connect

**Review by GuillaumeRx — APPROVED** [review]

**Review by GuillaumeRx — APPROVED** [review]

**Pull request merged**